File: cpp/5_Longest Palindromic Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int left, right, length, offset = 0, max = 0, st, ed;
        string result;
        for (int i = 0; i < s.length(); i++){
            // odd
            length = 1;
            offset = 1;
            if ((i - 1 >= 0) && (i + 1 < s.length())) {
                left = s[i - 1];
                right = s[i + 1];                
                while (left == right) {
                    offset += 1;
                    if ((i - offset >= 0) && (i + offset < s.length())) {                        
                        left = s[i - offset];
                        right = s[i + offset];                        
                    } else {
                        break;                    
                    }
                }
                length = 1 + 2 * (offset - 1);
            }
            if (max < length) {
                max = length;
                st = i - offset + 1;
                ed = i + offset - 1;
            }
            
            // even
            length = 0;
            offset = 1;
            if (i - 1 >= 0){
                left = s[i - 1];
                right = s[i];
                offset = 1;
                while (left == right){
                    offset += 1;
                    if ((i - offset >= 0) && (i + offset - 1 < s.length())) {
                        left = s[i - offset];
                        right = s[i + offset - 1];
                    }
                    else {
                        break;
                    }
                }
                length = 2 * (offset - 1);
            }
            if (max < length) {
                max = length;
                st = i - offset + 1;
                ed = i + offset - 2;
            }
        }
        result = s.substr(st, max);        
        return result;
    }
};

File: java/5_Longest Palindromic Substring.java
class Solution {
    public String longestPalindrome(String s) {
        int result = 0;
        int sl = 0;
        int sr = s.length() - 1;
        for (int i = 0; i < s.length(); ++i) {
            int temp = 1;
            int l = i - 1;
            int r = i + 1;
            while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
                temp += 2;
                l -= 1;
                r += 1;
            }
            if (result < temp) {
                result = temp;
                sl = l + 1;
                sr = r - 1;
            }
            
            temp = 0;
            l = i;
            r = i + 1;
            while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
                temp += 2;
                l -= 1;
                r += 1;
            }
            if (result < temp) {
                result = temp;
                sl = l + 1;
                sr = r - 1;
            }
        }
        return s.substring(sl, sr + 1);
    }
}